weed/filer/postgres3/postgres3_store.go

```diff
@@ -0,0 +1,512 @@
+package postgres3
+
+/*
+ * Copyright 2022 Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/lib/pq"
+	_ "github.com/lib/pq"
+	"github.com/seaweedfs/seaweedfs/weed/filer"
+	"github.com/seaweedfs/seaweedfs/weed/filer/abstract_sql"
+	"github.com/seaweedfs/seaweedfs/weed/glog"
+	"github.com/seaweedfs/seaweedfs/weed/pb/filer_pb"
+	"github.com/seaweedfs/seaweedfs/weed/util"
+)
+
+const (
+	CONNECTION_URL_PATTERN = "host=%s port=%d sslmode=%s connect_timeout=30"
+
+	createTablePattern = `CREATE TABLE IF NOT EXISTS "%s" (
+		key       varchar(65535) PRIMARY KEY,
+		name      varchar(65535),
+		prefixes  bigint[],
+		meta      bytea
+	)`
+	createTableIndexPattern = `CREATE INDEX on "%s" USING gin (prefixes);`
+	deleteTablePattern      = `DROP TABLE "%s";`
+	insertEntryPattern      = `INSERT INTO "%s" (key, name, prefixes, meta) VALUES ($1, $2, $3, $4)
+	ON CONFLICT (key)
+	DO
+	  UPDATE SET meta = EXCLUDED.meta;`
+	findEntryPattern      = `SELECT meta FROM "%s" WHERE key = $1`
+	deleteEntryPattern    = `DELETE FROM "%s" WHERE key = $1`
+	listEntryQueryPattern = `SELECT key, name, isdir, meta FROM
+	(
+		SELECT key, name, false as isdir, meta FROM "%s"
+		WHERE prefixes @> $1 AND cardinality(prefixes) < $2 - 1
+		AND name __COMPARISON__ $3 AND name LIKE $4 ORDER BY key ASC LIMIT $6
+	) s1
+	UNION
+	(
+		SELECT dir, dir, true isdir, NULL::bytea meta FROM
+		(
+			SELECT DISTINCT split_part(key, '/', $2) AS dir FROM "%s"
```

Comment:
Is this correct? If a key is `/a/b/c`, `split_part(key, '/', $2)` would return `a` as dir?

Reply:
Yes I think it should be correct because `split_part` starts counting at 1, not 0, but let me double check.

Reply:
I haven't had a chance to confirm this `split_part` behavior, may need a couple of days before I can.